BayeuxClient.agent.lib.nut

```diff
@@ -0,0 +1,936 @@
+const BAYEUX_CLIENT_META_CHANNEL_UNSUBSCRIBE            = "/meta/unsubscribe";
+
+class Bayeux {
+    static VERSION = "1.0.0";
+}
+
+class Bayeux.Client {
+    // TODO: Make disabled by default
+    _debugEnabled           = true;
+
+    _isDisconnected         = true;
+    _isDisconnecting        = false;
+    _isConnected            = false;
+    _isConnecting           = false;
+
+    _onConnectedCb          = null;
+    _onDisconnectedCb       = null;
+
+    _configuration          = null;
+    _advice                 = null;
+    _backoff                = 0;
+    _curMsgId               = 1;
+    _clientId               = null;
+    _cookies                = null;
+
+    // Topic -> handler for user messages (events)
+    _userHandlers           = null;
+
+    // Id -> handler for meta messages
+    _metaHandlers           = null;
+
+    // A timer for delayed connection
+    _reconnectTimer         = null;
+    // Bayeux protocol recommeds to keep at most 2 open HTTP requests from client to server.
+    // So we have "connect" and "extra" requests.
+    // The "connect" (heartbeat) request is open and waiting for a response almost all the time.
+    _connectHttpRequest     = null;
+    // A dedicated HTTP request to a server.
+    // The "extra" request allows to send messages without waiting the "connect" request to be answered.
+    // Array [<httprequest>, <callback>].
+    _extraHttpRequest       = null;
+    // Array of arrays [<message>, <callback>]. Here we keep messages waiting the "extra" request to be free.
+    _messageQueue           = null;
+
+
+    // Bayeux Client class constructor.
+    //
+    // Parameters:
+    //     configuration : Table        Key-value table with settings. There are required and optional settings.
+    //     onConnected : Function       Callback called every time the client is connected.
+    //          (optional)              The callback signature:
+    //                                  onConnected(error), where
+    //                                      error :             null if the connection is successful, error details otherwise.
+    //                                          Bayeux.Error
+    //     onDisconnected : Function    Callback called every time the client is disconnected.
+    //          (optional)              The callback signature:
+    //                                  onDisconnected(reason), where
+    //                                      reason :            null if the disconnection was caused by the disconnect() method,
+    //                                          Bayeux.Error    error details which explains a reason of the disconnection otherwise.
+    //
+    // Returns:                         Bayeux.Client instance created.
+    constructor(configuration, onConnected = null, onDisconnected = null) {
+        const REQUEST_INDEX = 0;
+        const CALLBACK_INDEX = 1;
+        const MESSAGE_INDEX = 1;
```

Comment:
Should this be equal to 2??

Reply:
If so, please add some tests to cover this.

Reply:
No, it should be 1

Reply:
Then I don't understand `send([msgCb[MESSAGE_INDEX]], msgCb[CALLBACK_INDEX]);`. this statement was what initially concerned me.

Reply:
Yes, that's a bug! MESSAGE_INDEX should be 0. Fixed that and added a test.